[MRG] Combining LOF and Isolation benchmarks

MaiRajborirug opened the pull request:
**What does this implement/fix? Explain your changes.**
Create one example that combines both LOF and IF ROC curves, according to the last discussion in a merged PR https://github.com/scikit-learn/scikit-learn/pull/9798#issuecomment-339322399.

**Other Comments**
Also fixed some data input typos in the `plot_annomaly_comparison.py`

MaiRajborirug commented:
![figure](https://user-images.githubusercontent.com/54995188/75651592-d5944c80-5c26-11ea-9ea3-14fdfb0e4310.png)

jnothman commented on examples/plot_anomaly_comparison.py:
```diff
@@ -2,30 +2,25 @@
 This example shows characteristics of different anomaly detection algorithms
 on 2D datasets. Datasets contain one or two modes (regions of high density)
 to illustrate the ability of algorithms to cope with multimodal data.
-
```
I don't think the changes to this file are warranted. Better to stick to one goal in the pull request. Please revert changes to this file.

MaiRajborirug replied:
Sure, got it. Thank you.

jnothman reviewed (commented):
Should this PR be removing bench_lof.py and/or bench_isolation_forest.py?

jnothman commented:
> Should this PR be removing bench_lof.py and/or bench_isolation_forest.py?

Please let us know when you've addressed this, or if you need help.

Currently this appears to have some lining errors

MaiRajborirug commented:
> > Should this PR be removing bench_lof.py and/or bench_isolation_forest.py?
> 
> Please let us know when you've addressed this, or if you need help.
> 
> Currently this appears to have some lining errors

Sorry about that. I just delete them.

albertcthomas commented:
The goal is to merge these two benchmarks into an example. Quoting @ogrisel's [comment](https://github.com/scikit-learn/scikit-learn/pull/9798#issuecomment-339322399)
> I had a quick IRL conversation with @albertcthomas and I think we should convert those 2 benchmark scripts into a single example script with a plt.subplot grid, one subplot per dataset with both LOF and IF ROC curves.

@MaiRajborirug can you please move the merged benchmark to examples/ and change the name of script to match the usual names of the examples?

MaiRajborirug commented:
@albertcthomas, sure. Is the name `plot_anomaly_bench.py` good?

albertcthomas commented:
> @albertcthomas, sure. Is the name `plot_anomaly_bench.py` good?

Looks good to me. I will try to review your PR soon.

MaiRajborirug commented:
@albertcthomas, just a followup comment. How was the review going?

albertcthomas reviewed (commented):
A first pass. This example will need to be referred to in the documentation on Novelty and Outlier detection.

albertcthomas commented on examples/plot_anomaly_bench.py:
```diff
@@ -1,11 +1,12 @@
+Comparison of LocalOutlierFactor and IsolationForest performances on 
+classical anomaly detection datasets using ROC curves
+====================================================================
 Benchmark plots of LocalOutlierFactor (LOF) and IsolationForest (IF) on 
```
Sorry I meant changing the sentence "Benchmark plots of LocalOutlierFactor..." not the title...

albertcthomas commented:
You also need to make the tests pass: you have some flake8 errors that you need to fix.

MaiRajborirug commented:
> You also need to make the tests pass: you have some flake8 errors that you need to fix.
Fixed it

> A first pass. This example will need to be referred to in the documentation on Novelty and Outlier detection.
Could you tell me where is the document [`Novelty and Outlier detection`](https://scikit-learn.org/stable/_sources/modules/outlier_detection.rst.txt) link in Github? I only found it in the Sklearn website.

MaiRajborirug commented:
> You also need to make the tests pass: you have some flake8 errors that you need to fix.

Fixed it. Thank you.

> A first pass. This example will need to be referred to in the documentation on Novelty and Outlier detection.

Do you know where is the document [`Novelty and Outlier detection`](https://scikit-learn.org/stable/_sources/modules/outlier_detection.rst.txt) link in Github? I only found it in the Sklearn website. And should I adjust the document file as well?

albertcthomas commented on examples/plot_anomaly_bench.py:
```diff
@@ -1,19 +1,18 @@
+===================================
+Comparison of LocalOutlierFactor and IsolationForest performances on
+classical anomaly detection datasets using ROC curves. Note that
+LocalOutlierFactor is not meant to predict on a test set by default and its
```
in fact I am just realizing now that we don't need the sentence about LOF here. However we should say that it is an outlier detection benchmark and keep the evaluation description.

albertcthomas reviewed (commented):
The documentation is located here: https://github.com/scikit-learn/scikit-learn/blob/master/doc/modules/outlier_detection.rst

MaiRajborirug commented:
> The documentation is located here: https://github.com/scikit-learn/scikit-learn/blob/master/doc/modules/outlier_detection.rst

I might need some help with inserting new `.png` file into the document. I can't find the folder `../auto_examples/ensemble/images/`

albertcthomas commented on doc/modules/.ipynb_checkpoints/outlier_detection-checkpoint.rst:
```diff
@@ -0,0 +1,395 @@
+.. _outlier_detection:
```
this file should not be committed, please remove it

albertcthomas commented on doc/modules/outlier_detection.rst:
```diff
@@ -112,14 +107,33 @@ an ellipse. For more details on the different estimators refer to the example
+.. figure:: ../auto_examples/images/sphx_glr_plot_anomaly_comparison_001.png
+   :target: ../auto_examples/plot_anomaly_comparison.html
+   :align: center
+   :scale: 50
```
why did you change the location of this figure? IMO it should stay at the same position.

albertcthomas commented on doc/modules/outlier_detection.rst:
```diff
@@ -112,14 +107,33 @@ an ellipse. For more details on the different estimators refer to the example
+:class:`neighbors.LocalOutlierFactor` because LOF is not meant to predict on
+a test set by default.
+
+.. figure:: sphx_glr_plot_anomaly_bench_001.png
```
I think it should be something like this. The png will be generated by sphinx when building the doc
```suggestion
.. figure:: ../auto_examples/images/sphx_glr_plot_anomaly_bench_001.png
```

albertcthomas replied:
actually I am not sure that we want to put the figure in the User Guide. I would remove it and just leave what you added in the Examples section.

albertcthomas commented on doc/modules/outlier_detection.rst:
```diff
@@ -112,14 +107,33 @@ an ellipse. For more details on the different estimators refer to the example
+The following figure benchmarks outlier detection algorithms using ROC
+(Receiver Operating Characteristics) curve on fives different datasets. 
+:class:`ensemble.IsolationForest` outperforms
+:class:`neighbors.LocalOutlierFactor` because LOF is not meant to predict on
```
this is not true, please remove this sentence.

albertcthomas commented on doc/modules/outlier_detection.rst:
```diff
@@ -112,14 +107,33 @@ an ellipse. For more details on the different estimators refer to the example
     :class:`covariance.EllipticEnvelope`.
-
+    
+  * See :ref:`sphx_glr_auto_examples_plot_anomaly_bench.py` for an anomaly
```
See ... for an example showing how to evaluate outlier detection estimators using ROC curves.

albertcthomas commented on examples/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,119 @@
+
+This example benchmarks outlier detection algorithms using ROC curves.
+The data are from classical anomaly detection datasets. Note that one of the
+algorithm, LocalOutlierFactor, is not meant to predict on a test set by default
```
please remove this sentence about LocalOutlierFactor.

albertcthomas reviewed (commented):
Thanks for you work @MaiRajborirug. There is a few things to fix/do in order for this to be good on my side :).

MaiRajborirug commented:
@albertcthomas does the changes look good? Please let me know if you want to adjust the document any further.

albertcthomas commented on examples/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,118 @@
+Anomaly detection's ROC benchmarks
+==================================
+
+This example benchmarks outlier detection algorithms using ROC curves.
```
This example benchmarks outlier detection algorithms using ROC curves on classical anomaly detection datasets. The algorithm performance is assessed in an outlier detection context:

albertcthomas commented on examples/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,118 @@
+2. The ROC curve is computed on the same dataset using the knowledge of the
+labels.
+
+This experiment uses area under the ROC curve score (AUC) to compare
```
I would remove this sentence as it is already said above that we are using ROC curves.

MaiRajborirug commented:
@albertcthomas, should we discuss the algorithm's performance in the example or not?

albertcthomas commented:
> @albertcthomas, should we discuss the algorithm's performance in the example or not?

Yes. I would discuss how to read the curves, that we want a higher ROC curve means a better performance and that we are usually mostly interested in low values of the FPR. I am not sure that we really want to emphasize any difference in performance between Isolation Forest and LOF. IMO this example is to show how to compare two outlier detection estimators.

MaiRajborirug commented on examples/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,124 @@
+is at low value of the false positive rate (FPR). The better algorithm
+have the curve on the top-left of the plot and the area under curve (AUC)
+close to 1. The diagonal dash line represnts a total random classification
+of outliers and inliers.
```
Is this additional section clear about interpreting ROC?

MaiRajborirug replied:
Is this section about `Interpreting the ROC plot` clear?

MaiRajborirug reviewed (commented):
I added the section about interpreting the ROC curve. @albertcthomas, is it short and clear enough?

albertcthomas commented on examples/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,129 @@
+The algorithm performance relates to how good the true positive rate (TPR)
+is at low value of the false positive rate (FPR). The better algorithm
+have the curve on the top-left of the plot and the area under curve (AUC)
+close to 1. The diagonal dash line represnts a total random classification
```
```suggestion
close to 1. The diagonal dashed line represents a random classification
```

albertcthomas commented on examples/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,129 @@
+# outlier detection models
+models = [
+    ("LOF", LocalOutlierFactor(n_neighbors=20, contamination="auto")),
+    ("IF", IsolationForest(n_jobs=-1, random_state=random_state,
```
do not use n_jobs=-1, always use n_jobs=1 by default

MaiRajborirug commented:
Thank you, I just updated the code as you suggested @albertcthomas .

albertcthomas commented:
You can see the rendered example [here](https://98044-843222-gh.circle-artifacts.com/0/doc/auto_examples/plot_anomaly_bench.html). Could you improve the rendering of the legends in the plot so that they can fit in the figure and do not have `'` in their texts?

MaiRajborirug commented:
> You can see the rendered example [here](https://98044-843222-gh.circle-artifacts.com/0/doc/auto_examples/plot_anomaly_bench.html). Could you improve the rendering of the legends in the plot so that they can fit in the figure and do not have `'` in their texts?

I fixed it. Also, I changed the legend position from `(loc="lower_right")` to `(loc="best")` so that the legend will not  block the y_axis and the plot

MaiRajborirug commented:
> You can see the rendered example [here](https://98044-843222-gh.circle-artifacts.com/0/doc/auto_examples/plot_anomaly_bench.html). Could you improve the rendering of the legends in the plot so that they can fit in the figure and do not have `'` in their texts?

I fixed the `'` and shortened the plot label description

MaiRajborirug commented on examples/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,129 @@
+        fpr, tpr, thresholds = roc_curve(y, scoring)
+        AUC = auc(fpr, tpr)
+        label_ = str(model_name
+                     + ": ROC for %s (AUC = %0.3f, \u0394t: %0.2fs)"
```
Shorten the code by changing from `train time` to `Δt`

albertcthomas replied:
I think that `train time` was better as it is not really a delta t here

MaiRajborirug commented:
@albertcthomas do you think this example is ready?

albertcthomas commented:
It is also a bit sad that LOF is not performing well on these datasets

MaiRajborirug commented:
I agree that LOF doesn't perform well here. @albertcthomas do you have any further suggestions on this PR improvement?

MaiRajborirug commented:
@albertcthomas do you think this example should be in the sklearn yet?

MaiRajborirug commented:
I had an error on [ci/circleci: doc-min-dependencies](https://circleci.com/gh/scikit-learn/scikit-learn/102299?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link). Could someone take a look?

The following code shows an error description. I checked my python version and it is 3.7.3, but the test said it is 3.5.6
```
+ python setup.py build_ext --inplace -j 3
Partial import of sklearn during the build process.
Traceback (most recent call last):
  File "setup.py", line 304, in <module>
    setup_package()
  File "setup.py", line 290, in setup_package
    % (platform.python_version(), sys.executable))
RuntimeError: Scikit-learn requires Python 3.6 or later. The current Python version is 3.5.6 installed in /home/circleci/miniconda/envs/testenv/bin/python.

Exited with code exit status 1
```

MaiRajborirug commented:
I had an error on [ci/circleci: doc-min-dependencies](https://circleci.com/gh/scikit-learn/scikit-learn/102299?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link). Could someone take a look?

The following code shows an error description. I checked my python version and it is 3.7.3, but the test said it is 3.5.6
```
+ python setup.py build_ext --inplace -j 3
Partial import of sklearn during the build process.
Traceback (most recent call last):
  File "setup.py", line 304, in <module>
    setup_package()
  File "setup.py", line 290, in setup_package
    % (platform.python_version(), sys.executable))
RuntimeError: Scikit-learn requires Python 3.6 or later. The current Python version is 3.5.6 installed in /home/circleci/miniconda/envs/testenv/bin/python.

Exited with code exit status 1
```

MaiRajborirug commented:
Seem like my repo doesn't get updated from the `sckit-learn: master`. I instead created a [#PR17147](https://github.com/scikit-learn/scikit-learn/pull/17147) which is up to date and pass the tests. Please continue the discussion that PR.

Commit:
add bench files

MaiRajborirug commented:
@albertcthomas now the code passed the tests, could I ask for a full review?

albertcthomas commented:
Thanks for fixing the tests @MaiRajborirug. I will not be able to do a review before beginning of June. One thing that I would like to try to improve is to find a way to make the example a bit more "beautiful". I don't know if changing the parameters can improve the performance of LOF a bit of if we could find one additional not too large dataset (because the example still needs to be quick to run).

MaiRajborirug commented:
Thank you so much! I can try adding new dataset and adjusting parameters before June then

Commit:
add more outliers datasets

Commit:
update python file

Commit:
shorten the .py file

MaiRajborirug commented:
I added the outliers datasets from http://odds.cs.stonybrook.edu/ into the code.
Here is the updated preview figure:
![image](https://user-images.githubusercontent.com/54995188/83199600-8e2b3880-a10f-11ea-9435-66308dd1d69d.png)

Pimpwhippa commented:
@MaiRajborirug 
> I had an error on ci/circleci: doc-min-dependencies. Could someone take a look?
Seem like my repo doesn't get updated from the sckit-learn: master

Did you just pull from master onto this open PR, and that made it passed the test? I have the same error. Wondering what is the best can I do. Thanks.

MaiRajborirug commented:
> Did you just pull from master onto this open PR, and that made it passed the test? I have the same error. Wondering what is the best can I do. Thanks.

Hi @Pimpwhippa. My error happened because a PR has opened for so long that the other files in scikit-learn already updated. Thus, there is [merge conflict](https://dont-be-afraid-to-commit.readthedocs.io/en/latest/git/conflicts.html). The problem continues because I cannot simply delete my forked repo if I have already created a PR.

Here are what I did:
Note: you can try `git fetch upstream` Unfortunately, I didn't work for me.
1. download a master repo `scikit-learn/scikit-learn` into my local folder1
2. download a forked repo `MaiRajborirug/scikit-learn` into my local folder2
3. keep the files I want to PR into folder3
3. paste folder1 to folder2 (folder2 will be just like the recent sckit-learn forked repo)
4. add, update, delete files you want to PR from folder3 to folder2
5. `git push` folder2 

Hope this help

Pimpwhippa commented:
@MaiRajborirug Thank you for your suggestion. I tried pull and then rebase, but the result is still ugly. I got all other changes in my PR. Maybe I'll try your method if this doesn't work.

albertcthomas commented on examples/miscellaneous/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,148 @@
+    if dataset_name == "http" or dataset_name == "smtp":
+        y = (y != b"normal.").astype(int)
+
+    if dataset_name in ["glass", "vowels", "wbc", "letter"]:
```
I am not sure about the policy regarding pushing datasets to the sklearn repo. Please fetch the datasets from [open ml](https://www.openml.org/home) using the `sklearn.datasets.fetch_openml` function (at least glass and letter are available).

albertcthomas reviewed (commented):
Thanks for the new datasets. They show that LOF can sometimes work better than Iforest as well as offer a better visualization, even if this is not the major goal of a benchmark.

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

Commit:
remove the datasets

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

update the forked repo

albertcthomas commented:
Thanks @MaiRajborirug. This is starting to look much better now :). I will try to do a full review soon. Sorry for the delay and thanks for your patience.

MaiRajborirug commented:
@albertcthomas thank you for your suggestion! Let me know if you have any suggestions.

Commit:
update the comment

Commit:
update doc

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

MaiRajborirug commented:
@albertcthomas so sorry for the late update. I think this PR is ready for the full review if you have time.

Commit:
update doc2

Commit:
update doc2

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

MaiRajborirug commented:
@albertcthomas can you review this PR one more time?

MaiRajborirug commented:
@albertcthomas  @jnothman could you check if there is anything that needs to adjust in this PR?

albertcthomas commented:
Thanks for the reminders and sorry for the delay. I will try to have a look
soon.

On Wed 26 Aug 2020 at 15:01, Mai <notifications@github.com> wrote:

>
>
> @albertcthomas <https://github.com/albertcthomas> @jnothman
> <https://github.com/jnothman> could you check if there is anything that
> needs to adjust in this PR?
>
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/16606#issuecomment-680864081>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADZ2DLWRDWNNEWAQW4EWCZ3SCUBRNANCNFSM4K7MKNDQ>
> .
>
>
>

albertcthomas commented on examples/miscellaneous/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,160 @@
+Interpreting the ROC plot
+-------------------------
+The algorithm performance relates to how good the true positive rate (TPR)
+is at low value of the false positive rate (FPR). The better algorithm
```
algorithms

albertcthomas commented on benchmarks/bench_isolation_forest.py:
```diff
@@ -1,159 +0,0 @@
-"""
```
I don't think this file should be removed. This benchmark is done in the novelty context which is not the case of the example

MaiRajborirug replied:
Got it

albertcthomas commented on benchmarks/bench_lof.py:
```diff
@@ -1,106 +0,0 @@
-outliers.
-2. The ROC curve is computed on the same dataset using the knowledge of the
-labels.
-In this context there is no need to shuffle the dataset because the model
```
please add this to the new example.

MaiRajborirug replied:
Got it.

albertcthomas replied:
This file should be removed. `bench_isolation_forest.py` should be kept as it is a novelty detection benchmark but `bench_lof.py` should be removed as it is replaced by the new example.

albertcthomas commented on examples/miscellaneous/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,160 @@
+        X = dataset.data
+        y = dataset.target
+        if dataset_name == "SA":
+            idx = np.random.choice(
```
why is this needed? also if we need it you should use `random_state`

MaiRajborirug replied:
Sure I can add `np.random.seed(random_state)`. For your question, `SA` has very high dimensions (~41) comparing to `http`, `smtp`, and `SF`. I picked fewer samples to reduce run time.

MaiRajborirug replied:
Solved

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
add the examples

Commit:
add examples ver.2

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
debugged

MaiRajborirug commented:
Hi, @albertcthomas I adjusted the code according to your advice. Still, I used `np.rando,m.choice`  on `SA` to further downsample to reduce run time.

Commit:
Merge remote-tracking branch 'upstream/master'

MaiRajborirug commented:
@albertcthomas, have you looked at this update yet? If so, please let me know your oppinion about it.

albertcthomas commented on examples/miscellaneous/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,158 @@
+    plt.subplot(len(datasets), 1, dataset_idx + 1)
+
+    # loading and vectorization
+    print("loading data: ", str(dataset_idx + 1))
```
```suggestion
    print(f"Loading {dataset_name} data")
```

albertcthomas commented on examples/miscellaneous/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,158 @@
+        if model_name == "LOF":
+            scoring = -model.negative_outlier_factor_
+            # the lower score, the more normal
+        if model_name == "IF":
```
```suggestion
        if model_name == "IForest":
```

albertcthomas reviewed (changes requested):
I tried to run the example locally, please see my comments. I'll try to iterate faster in the next few days so that we can finish this PR. Thanks again

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
change the file name

Commit:
fix  error

MaiRajborirug commented on examples/miscellaneous/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,158 @@
+
+    if dataset_name in ["glass", "wdbc", "cardiotocography"]:
+        dataset = fetch_openml(name=dataset_name, version=1)
+        X = dataset.data.values  # from dataframe to array
```
I put `as_frame=False` in the line 107 to make sure that the outputs are `numpy.ndarray` not `DataFrames`. It should work now.

albertcthomas commented on examples/miscellaneous/plot_anomaly_bench.py:
```diff
@@ -0,0 +1,158 @@
+
+    # loading and vectorization
+    print("loading data: ", str(dataset_idx + 1))
+    if dataset_name in ["http", "smtp", "SA", "SF"]:
```
my bad, you're right

Commit:
update doc/modules/outlier_detection.rst

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,160 @@
+        dataset = fetch_covtype()
+        X = dataset.data
+        y = dataset.target
+        idx = rng.choice(X.shape[0],
+                         int(X.shape[0]*0.1), replace=False)
```
```suggestion
        idx = rng.choice(X.shape[0], int(X.shape[0]*0.1), replace=False)
```

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,160 @@
+    print("vectorizing data")
+
+    if dataset_name == "SF":
+        idx = rng.choice(X.shape[0],
+                         int(X.shape[0]*0.1), replace=False)
```
```suggestion
        idx = rng.choice(X.shape[0], int(X.shape[0]*0.1), replace=False)
```

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,160 @@
+        y = (y != b"normal.").astype(int)
+
+    if dataset_name == "SA":
+        idx = rng.choice(X.shape[0],
+                         int(X.shape[0]*0.1), replace=False)
```
```suggestion
        idx = rng.choice(X.shape[0], int(X.shape[0]*0.1), replace=False)
```

albertcthomas reviewed (commented):
We need to fix the plot rendering.

Commit:
plot rendering

MaiRajborirug commented:
> We need to fix the plot rendering.

Just fixed them

albertcthomas commented:
Rendering is [here](https://121483-843222-gh.circle-artifacts.com/0/doc/auto_examples/miscellaneous/plot_outlier_detection_bench.html). There is too much space between the figure and the text. The legend is sometimes too big for the plot. I would put the dataset name as the title of each subplot and remove it from the legend. Maybe do a grid of 2x4.

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,157 @@
+
+        fpr, tpr, thresholds = roc_curve(y, scoring)
+        AUC = auc(fpr, tpr)
+        label_ = str(model_name
+                     + ": %s (AUC = %0.3f, train time: %0.2fs)"
+                     % (dataset_name, AUC, fit_time))
```
```suggestion
        label_ = (f'{model_name} (AUC = {area:0.3f}, '
                  f'train time: {fit_time:0.2f})')
```

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,157 @@
+            scoring = -model.fit(X).decision_function(X)
+
+        fpr, tpr, thresholds = roc_curve(y, scoring)
+        AUC = auc(fpr, tpr)
```
```suggestion
        area = auc(fpr, tpr)
```

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,157 @@
+    plt.xlim([-0.05, 1.05])
+    plt.ylim([-0.05, 1.05])
+    plt.legend(loc="lower right")
+    if dataset_idx == 0:
+        plt.title("Receiver operating characteristic (ROC)")
```
```suggestion
    plt.title(dataset_name)
```

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
adjust plots

MaiRajborirug commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,157 @@
+    ("IForest", IsolationForest(random_state=rng, contamination="auto")),
+]
+
+rows = math.ceil(len(datasets)/2)
+plt.figure(figsize=(10, rows * 3))
+for dataset_idx, dataset_name in enumerate(datasets):
+    plt.subplot(rows, 2, dataset_idx + 1)
```
Changed from 1 to 2 columns plots

MaiRajborirug commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,157 @@
+    if dataset_idx == len(datasets) - 1:
+        plt.xlabel("False Positive Rate")
+        plt.ylabel("True Positive Rate")
+plt.tight_layout(pad=2.0)  # spacing between subplots
```
added some spaces between plots for clear titles

MaiRajborirug reviewed (commented):
Here is the update in plot rendering.
![image](https://user-images.githubusercontent.com/54995188/95143074-3054d000-0743-11eb-8f9c-47b520e6783e.png)

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,157 @@
+        fit_time = time() - tstart
+        if model_name == "LOF":
+            scoring = -model.negative_outlier_factor_
+            # the lower score, the more normal
```
remove this comment add before `if model_name == "LOF":`
```python
        # use oppositve of default scoring so that the lower score, the more
        # normal
```

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,157 @@
+        X = dataset.data
+        y = dataset.target
+
+    if dataset_name == "forestcover":
```
please put this after all the kdd cup datasets so that kdd cup dataset preprocessing follows fetch_kddcup99

MaiRajborirug replied:
Can you clarify what do you mean in this comment? Do I add `fetch_kddcup99(subset=None)` or do I put 
```
x = dataset.data
y = dataset.target
```
below all `if` in the loading data section?

MaiRajborirug replied:
Or should we merge loading and vectorizing to be one section?

MaiRajborirug replied:
Now I got it. My change is in the next commit

albertcthomas replied:
sorry this was not very clear

albertcthomas reviewed (commented):
Last comments.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
update .py

MaiRajborirug commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,159 @@
+        X = np.c_[X[:, :1], x1, x2, x3, X[:, 4:]]
+        y = (y != b"normal.").astype(int)
+
+    if dataset_name == "http" or dataset_name == "smtp":
+        y = (y != b"normal.").astype(int)
+
+    if dataset_name == "forestcover":
```
now `if dataset_name == "forestcover":` is under the whole `fetch_kddcup99` data

MaiRajborirug commented:
@albertcthomas, thank you so much for going through this PR with me from the start!!

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,159 @@
+    print("vectorizing data")
+    if dataset_name == "SF":
+        idx = rng.choice(X.shape[0], int(X.shape[0]*0.1), replace=False)
+        X = X[idx, :]  # reduce the sample size
```
```suggestion
        X = X[idx]  # reduce the sample size to speed up computation
```

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,159 @@
+
+    if dataset_name == "SA":
+        idx = rng.choice(X.shape[0], int(X.shape[0]*0.1), replace=False)
+        X = X[idx, :]  # reduce the sample size
```
```suggestion
        X = X[idx]  # reduce the sample size to speed up computation
```

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,159 @@
+        X = dataset.data
+        y = dataset.target
+        idx = rng.choice(X.shape[0], int(X.shape[0]*0.1), replace=False)
+        X = X[idx, :]
```
X = X[idx]  # reduce the sample size to speed up computation

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,159 @@
+            X_ben, y_ben = X[~s], y[~s]
+
+            # downsampled to 39 points (9.8% outliers)
+            idx = rng.choice(
+                y_mal.shape[0], 39, replace=False)
```
```suggestion
            idx = rng.choice(y_mal.shape[0], 39, replace=False)
```

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,159 @@
+            X_mal, y_mal = X[s], y[s]
+            X_ben, y_ben = X[~s], y[~s]
+
+            # downsampled to 39 points (9.8% outliers)
```
```suggestion
            # downsample to 39 points (9.8% outliers) to speed up computation
```

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,159 @@
+            # downsampled to 39 points (9.8% outliers)
+            idx = rng.choice(
+                y_mal.shape[0], 39, replace=False)
+            X_mal2 = X_mal[idx, :]
```
```suggestion
            X_mal2 = X_mal[idx]
```

albertcthomas commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,159 @@
+    plt.ylim([-0.05, 1.05])
+    plt.legend(loc="lower right")
+    plt.title(dataset_name)
+    if dataset_idx == len(datasets) - 1:
+        plt.xlabel("False Positive Rate")
+        plt.ylabel("True Positive Rate")
```
```suggestion
    plt.xlabel("False Positive Rate")
    plt.ylabel("True Positive Rate")
```

albertcthomas reviewed (approved):
Some final comments. Otherwise looks good to me! Thanks a lot @MaiRajborirug! This required a lot of work, including finding new datasets. Example is rendered [here](https://121568-843222-gh.circle-artifacts.com/0/doc/auto_examples/miscellaneous/plot_outlier_detection_bench.html).

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
update plot's axis

MaiRajborirug commented:
The latest example is [here](https://121616-843222-gh.circle-artifacts.com/0/doc/auto_examples/miscellaneous/plot_outlier_detection_bench.html).

MaiRajborirug commented:
@albertcthomas, is there anything left to, or can it be merged?

fufuengsin commented:
A useful example for illustrating performance evaluation. Awesome effort. LGTM!

MaiRajborirug commented:
@jnothman, could you take a look at this PR and see if there is anything to change or add?

albertcthomas commented:
> @albertcthomas, is there anything left to, or can it be merged?

we'll have to wait for other reviewers now, a PR must be approved by 2 core devs.

lorentzenchr commented:
@MaiRajborirug I see that this PR is waiting for a 2nd review for quite some time now. For a review, it usually helps if CI is green. Here, I see 2 failing tests.
Could you trigger CI by an empty commit and fix those CI failures in case they are still there?

MaiRajborirug committed:
Merge remote-tracking branch 'upstream/main'

MaiRajborirug commented:
Hi @lorentzenchr @albertcthomas @jnothman, I push the code again and found the error from `open_ml`. I can no longer access any dataset from [`open_ml`](https://www.openml.org/)

```
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate has expired (_ssl.c:1123)>

URLError                                  Traceback (most recent call last)
<ipython-input-15-443fd7a49542> in <module>
    102 
    103     if dataset_name in ["glass", "wdbc", "cardiotocography"]:
--> 104         dataset = fetch_openml(name=dataset_name, version=1,
    105                                as_frame=False)
    106         X = dataset.data
```

Any idea for to fix the problem? Or should I discard the `open_ml` datasets?

albertcthomas commented:
Hm do you have the issue locally?

MaiRajborirug commented:
I found that the SSL certificate on openml.org was expired recently. Here is the raising issue in open ML github https://github.com/openml/openml.org/issues/167

Here is the official webpage
![image](https://user-images.githubusercontent.com/54995188/112398589-ece32380-8cda-11eb-8d92-4dc5c92058b3.png)

Here is the error in from my local computer
![image](https://user-images.githubusercontent.com/54995188/112399091-f7ea8380-8cdb-11eb-908d-0e2910156978.png)

MaiRajborirug committed:
open_ml fixed the certificate issue

MaiRajborirug commented:
@albertcthomas @jnothman the PR should be good now. Could you review this one last time?

lorentzenchr commented:
> Some final comments. Otherwise looks good to me! 

@albertcthomas Would you like to approve as a reviewer (officially)?

albertcthomas reviewed (approved):
LGTM! I'm not sure if we want to keep `benchmarks/bench_lof.py` or not. I'd say that we can keep it as some datasets are downsampled for the sake of the example.

lorentzenchr commented:
@MaiRajborirug Could you mention"LOF" and "IForest" in the text together and say what these abbreviations stand for? Maybe also, why you picked those 2 algorithms.

MaiRajborirug committed:
Merge remote-tracking branch 'upstream/main'

MaiRajborirug committed:
update text

MaiRajborirug commented:
@lorentzenchr fixed it. By the way @albertcthomas, @jnothman hasn't been active sine Mar 2, 2020. If he is still busy, is it possible for me to ask for another reviewer to review my PR instead?

cmarmo commented:
A lot of work here from the contributor and the reviewers... perhaps @ogrisel might be interested in finalize this review (as per [his comment](https://github.com/scikit-learn/scikit-learn/pull/9798#issuecomment-339322399))?
@MaiRajborirug I know it has been a lot... perhaps you might find some time to synchronize with main? So the rendering of the example will be available. Thank you for your patience!

MaiRajborirug committed:
Merge remote-tracking branch 'upstream/main'

MaiRajborirug committed:
Merge remote-tracking branch 'upstream/main'

MaiRajborirug committed:
synchronize

MaiRajborirug committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn

MaiRajborirug committed:
up to date 2

MaiRajborirug committed:
up to date 3

MaiRajborirug committed:
up to date 4

MaiRajborirug committed:
up to date 5

MaiRajborirug commented:
@cmarmo thank you for suggesting, the PR has been updated!

MaiRajborirug commented:
@ogrisel could please you be my PR's second reviewer? I have already updated the code.

lorentzenchr commented:
@MaiRajborirug Thank you for the update and your contribution.

At the moment, I’m worried about having too many examples and there is already [this example](https://scikit-learn.org/stable/auto_examples/miscellaneous/plot_anomaly_comparison.html#sphx-glr-auto-examples-miscellaneous-plot-anomaly-comparison-py).

I propose to open an issue to discuss a possible example (new or embedded in an existing one) where outlier detection is an important part in an end to end prediction pipeline, in line what @glemaitre proposed in https://github.com/scikit-learn/scikit-learn/pull/16378#issuecomment-586188362:

> However, I agree with you that we miss an example where we should show an end-to-end pipeline where anomaly detection is beneficial in classification and this should be rigorously analyzed with such classification metrics/plots.

albertcthomas commented:
> At the moment, I’m worried about having too many examples and there is already [this example](https://scikit-learn.org/stable/auto_examples/miscellaneous/plot_anomaly_comparison.html#sphx-glr-auto-examples-miscellaneous-plot-anomaly-comparison-py).

The goal has been to transform the classical benchmark that is used to compare outlier detection estimators into an example for the users. This has been considered because with a given configuration the benchmark was fast to run. We don't have such an example in scikit-learn. The comparison one is only based on 2-d plots and is not using any metrics such as ROC curves or AUC.

I agree that what Guillaume suggests would be ideal and this could be discussed in an issue but I think the example of this PR has its place in scikit-learn.

MaiRajborirug committed:
Merge branch 'main' into master

MaiRajborirug commented:
The pull request says 'Merging is blocked' even though all checks have passed. Could someone tell me what happen?

cmarmo commented:
> The pull request says 'Merging is blocked' even though all checks have passed. Could someone tell me what happen?

The main branch is protected now. Only core-devs can merge... this is not really different from before for contributors but it is explicitly stated.
Thanks for synchronizing the pull request.
@ogrisel, @jeremiedbb, this is a very old PR with one approval needing a decision. Could it be possibly included in 1.1? Thanks!

cmarmo commented:
@MaiRajborirug this pull request has been added to the list for the [review sprint](https://mail.python.org/pipermail/scikit-learn/2022-April/004237.html). Feel free to join if you think it will speed up the process. Thanks.

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,164 @@
+2. The ROC curve is computed on the same dataset using the knowledge
+of the labels.
+
+There is no particular reason to choose algorithms LOF and ROC. The goal
```
```suggestion
There is no particular reason to choose algorithms LOF and IForest. The goal
```

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,164 @@
+Interpreting the ROC plot
+-------------------------
+The algorithm performance relates to how good the true positive rate (TPR)
+is at low value of the false positive rate (FPR). The better algorithms
```
```suggestion
is at low value of the false positive rate (FPR). The best algorithms
```

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,164 @@
+The algorithm performance relates to how good the true positive rate (TPR)
+is at low value of the false positive rate (FPR). The better algorithms
+have the curve on the top-left of the plot and the area under curve (AUC)
+close to 1. The diagonal dashed line represents a random classification
+of outliers and inliers.
```
There's no diagonal dashed line on the plots. Maybe you forgot to add it ?

jeremiedbb reviewed (commented):
Thanks for the PR @MaiRajborirug. We are currently moving our examples to use a notebook style (see https://github.com/scikit-learn/scikit-learn/issues/22406). Could you update your PR to follow these guidelines ? In particular, split the content in different cells with titles and move the imports in the cells where they are needed.

MaiRajborirug committed:
follow suggestions on notebook style

jeremiedbb committed:
Merge branch 'main' into master

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,186 @@
+# -----------------
+#
+# The example uses real world datasets available in `scikit-learn.datasets`.
+datasets = [
+    "http",
+    "smtp",
+    "SA",
+    "SF",
+    "forestcover",
+    "glass",
+    "wdbc",
+    "cardiotocography",
+]
```
I think it would be clearer if the data loading was actually in this section.

We could define a function preprocess_dataset that takes the dataset name. It would load the dataset and preprocess it and then returns X, y.

MaiRajborirug replied:
Currently, I loaded, preprocessed, and stored datasets lists Xs = [X1, X2,..] and Ys. It requires more memory but will finish all data preprocessing here. In the next code block, we will just call the data to compute ROV. Do you prefer this method?

jeremiedbb replied:
> It requires more memory

That's why I suggested to define a function to load and preprocess one dataset at a time. Then in the next cells we'll call the function for each dataset

MaiRajborirug replied:
Done it! I created 2 functions. one for the data processing and the other for the model prediction. The example will only call each function when needed.

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,186 @@
+
+print(__doc__)
+
+rng = np.random.RandomState(42)
```
I'd move that in the next cell

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,186 @@
+# Author: Pharuj Rajborirug <pharuj.ra@kmitl.ac.th>
+# License: BSD 3 clause
+
+import numpy as np
```
I'd move that in the next cell

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,186 @@
+    ("IForest", IsolationForest(random_state=rng, contamination="auto")),
+]
+
+# %%
+# Preprocess data and compute ROC curves
+# --------------------------------------
+#
+# During the data preprocessing, we reduce the sample size of some datasets to
+# speed up computation.
```
Following my previous comment, I think this cell would be clearer if it was just doing the plot.

MaiRajborirug committed:
Circleci solved

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,186 @@
+        if model_name == "IForest":
+            scoring = -model.fit(X).decision_function(X)
+
+        fpr, tpr, thresholds = roc_curve(y, scoring)
+        area = auc(fpr, tpr)
+        label_ = f"{model_name} (AUC = {area:0.3f}, train time: {fit_time:0.2f})"
+        plt.plot(fpr, tpr, lw=1, label=label_)
```
Let's use `RocCurveDisplay.from_estimator` (see https://scikit-learn.org/stable/modules/generated/sklearn.metrics.RocCurveDisplay.html#sklearn.metrics.RocCurveDisplay.from_estimator)

We can define the axis for the plot when creating the plt.figure and pass them.

MaiRajborirug replied:
Look like 'from_estimator' doesn't work on IsolationForest. I will use 'from_predictions' instead.

MaiRajborirug committed:
Apply RocCurveDisplay.from_estimator

MaiRajborirug committed:
fixed Linting

MaiRajborirug committed:
fix Linting2

MaiRajborirug committed:
fix linting3

MaiRajborirug committed:
fix linting4

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,201 @@
+# Compute ROC curves
+# ------------------
+ys_pred = []
+pos_label = 0  # mean 0 belongs to positive class
```
This is used in next cell

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,201 @@
+# %%
+# Compute ROC curves
+# ------------------
+ys_pred = []
+pos_label = 0  # mean 0 belongs to positive class
+
+print("Estimators processing...")
+for i, dataset_name in enumerate(datasets_name):
+
+    ys_pred_ = []
+    for model_name, clf in models:
+        clf.fit(Xs[i])
+        if model_name == "LOF":
+            y_pred = clf.negative_outlier_factor_
+
+        if model_name == "IForest":
+            y_pred = clf.fit(Xs[i]).decision_function(Xs[i])
+
+        ys_pred_.append(y_pred)
+    ys_pred.append(ys_pred_)
```
maybe make ys_pred a dictionary ``ys_pred["http"]["LOF"]`` ? it would be a bit more explicit

MaiRajborirug replied:
Since I created a compute_prediction function. I will no longer use `ys_pred`

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,201 @@
+    ("IForest", IsolationForest(random_state=rng, contamination="auto")),
+]
+
+# %%
+# Compute ROC curves
+# ------------------
```
It's more Compute predictions, no ?

jeremiedbb replied:
Actually I think Defining the models (prev cell) and this one could be merged in a single cell. The prev cell does basically nothing

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,201 @@
+            ys_pred[i][j],
+            pos_label=pos_label,
+            name=model_name,
+            lw=lw,
```
let's use a more explicit name
```suggestion
            linewidth=linewidth,
```

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,201 @@
+        x1 = lb.fit_transform(X[:, 1].astype(str))
+        x2 = lb.fit_transform(X[:, 2].astype(str))
+        x3 = lb.fit_transform(X[:, 3].astype(str))
+        X = np.c_[X[:, :1], x1, x2, x3, X[:, 4:]]
```
nitpick: more explicit is better :) let's use np.column_stack instead of np.c_

jeremiedbb reviewed (commented):
Here's another pass @MaiRajborirug

MaiRajborirug committed:
create 2 functions + more explicit

ArturoAmorQ commented on doc/modules/outlier_detection.rst:
```diff
@@ -131,6 +131,10 @@ sections hereunder.
+  * See :ref:`sphx_glr_auto_examples_miscellaneous_plot_outlier_detection_bench.py`
+    for an example showing how to evaluate outlier detection estimators
+    using ROC curves.
+
```
Could you please add a tag `.. _local_outlier_factor:` in (current) line 317 of this file i.e. above the corresponding subsection? (Similar to the tag `.. _isolation_forest:` in line 252). Then, call this reference at the beginning of the example.

ArturoAmorQ commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,193 @@
+        dataset = fetch_kddcup99(subset=dataset_name, percent10=True, random_state=rng)
+        X = dataset.data
+        y = dataset.target
+    if dataset_name == "SF":
+        idx = rng.choice(X.shape[0], int(X.shape[0] * 0.1), replace=False)
+        X = X[idx]  # reduce the sample size
+        y = y[idx]
+        lb = LabelBinarizer()
+        x1 = lb.fit_transform(X[:, 1].astype(str))
+        X = np.c_[X[:, :1], x1, X[:, 2:]]
+        y = (y != b"normal.").astype(int)
+    if dataset_name == "SA":
+        idx = rng.choice(X.shape[0], int(X.shape[0] * 0.1), replace=False)
+        X = X[idx]  # reduce the sample size
+        y = y[idx]
+        lb = LabelBinarizer()
+        x1 = lb.fit_transform(X[:, 1].astype(str))
+        x2 = lb.fit_transform(X[:, 2].astype(str))
+        x3 = lb.fit_transform(X[:, 3].astype(str))
+        X = np.column_stack((X[:, :1], x1, x2, x3, X[:, 4:]))
+        y = (y != b"normal.").astype(int)
+    if dataset_name == "http" or dataset_name == "smtp":
+        y = (y != b"normal.").astype(int)
```
Maybe you can factorize some lines of code by inserting these if statements inside the more general 
`if dataset_name in ["http", "smtp", "SA", "SF"]:`
statement, i.e similarly to what you do for the individual cases in your larger  
`if dataset_name in ["glass", "wdbc", "cardiotocography"]:`
statement.

MaiRajborirug replied:
@ArturoAmorQ  Could you check? I think it should be good now

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,193 @@
+rng = np.random.RandomState(42)
+
+
+def preprocess_dataset(dataset_name="http"):
```
```suggestion
def preprocess_dataset(dataset_name):
```

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,193 @@
+# ----------------------------------
+#
+# The example uses real-world datasets available in `scikit-learn.datasets`
+# and the sample size of some datasets to speed up computation.
```
```suggestion
# and the sample size of some datasets is reduced to speed up computation.
```

jeremiedbb commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,193 @@
+from sklearn.ensemble import IsolationForest
+
+
+def compute_prediction(X, model_name="LOF"):
```
```suggestion
def compute_prediction(X, model_name):
```

jeremiedbb reviewed (approved):
just a couple more nitpicks on my side but looks good otherwise once the comments from @ArturoAmorQ.

It could also be interesting to check if we can give some insights about what makes the performance different between the 2 algorithms for different datasets. @ogrisel had the intuition that IForest could be more performant of datasets with heterogeneous, while LOF might be better with numeric features and rotational invariance for instance.

MaiRajborirug committed:
Fix docs and shorten code

MaiRajborirug commented:
@jeremiedbb that is a very interesting topic to dive into, but let's work it in another PR. It might be better to use generated data for that experiment where we have full control of the datasets.

jeremiedbb commented on examples/neighbors/plot_lof_outlier_detection.py:
```diff
@@ -3,7 +3,10 @@
 =================================================
 
-The Local Outlier Factor (LOF) algorithm is an unsupervised anomaly detection
+An example using :class:`~sklearn.neighbors.LocalOutlierFactor` for outlier
+detection.
+
+The Local Outlier Factor (LOF) algorithm is an unsupervised outlier detection
```
Let's not introduce unrelated changes

```suggestion
The Local Outlier Factor (LOF) algorithm is an unsupervised anomaly detection
```

ArturoAmorQ commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,191 @@
+Evaluation of outlier detection estimators
+==========================================
+
+This example benchmarks outlier detection algorithms, Local Outlier
+Factor (LOF) and Isolation Forest (IForest), using ROC curves on
```
```suggestion
This example benchmarks outlier detection algorithms, :ref:`local_outlier_factor` (LOF) and :ref:`isolation_forest` (IForest), using ROC curves on
```

jeremiedbb committed:
Update examples/neighbors/plot_lof_outlier_detection.py

jeremiedbb committed:
Update examples/miscellaneous/plot_outlier_detection_bench.py

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

ArturoAmorQ commented on examples/miscellaneous/plot_outlier_detection_bench.py:
```diff
@@ -0,0 +1,191 @@
+
+    # loading and vectorization
+    print(f"Loading {dataset_name} data")
+    if dataset_name in ["http", "smtp", "SA", "SF"]:
+        dataset = fetch_kddcup99(subset=dataset_name, percent10=True, random_state=rng)
+        X = dataset.data
+        y = dataset.target
+        if dataset_name in ["http", "smtp"]:
+            y = (y != b"normal.").astype(int)
+        if dataset_name in ["SA", "SF"]:
+            idx = rng.choice(X.shape[0], int(X.shape[0] * 0.1), replace=False)
+            X = X[idx]  # reduce the sample size
+            y = y[idx]
+            lb = LabelBinarizer()
+            x1 = lb.fit_transform(X[:, 1].astype(str))
+            if dataset_name == "SA":
+                x2 = lb.fit_transform(X[:, 2].astype(str))
+                x3 = lb.fit_transform(X[:, 3].astype(str))
+                X = np.column_stack((X[:, :1], x1, x2, x3, X[:, 4:]))
+            if dataset_name == "SF":
+                X = np.column_stack((X[:, :1], x1, X[:, 2:]))
+            y = (y != b"normal.").astype(int)
```
```suggestion
    if dataset_name in ["http", "smtp", "SA", "SF"]:
        dataset = fetch_kddcup99(subset=dataset_name, percent10=True, random_state=rng)
        X = dataset.data
        y = dataset.target
        lb = LabelBinarizer()

        if dataset_name == "SF":
            idx = rng.choice(X.shape[0], int(X.shape[0] * 0.1), replace=False)
            X = X[idx]  # reduce the sample size
            y = y[idx]
            x1 = lb.fit_transform(X[:, 1].astype(str))
            X = np.c_[X[:, :1], x1, X[:, 2:]]

        elif dataset_name == "SA":
            idx = rng.choice(X.shape[0], int(X.shape[0] * 0.1), replace=False)
            X = X[idx]  # reduce the sample size
            y = y[idx]
            x1 = lb.fit_transform(X[:, 1].astype(str))
            x2 = lb.fit_transform(X[:, 2].astype(str))
            x3 = lb.fit_transform(X[:, 3].astype(str))
            X = np.c_[X[:, :1], x1, x2, x3, X[:, 4:]]

        y = (y != b"normal.").astype(int)
```

This is what I had in mind

jeremiedbb committed:
Update examples/miscellaneous/plot_outlier_detection_bench.py

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

MaiRajborirug committed:
revert doc + shorten code

MaiRajborirug commented:
@jeremiedbb @ArturoAmorQ how were the changes?

ArturoAmorQ reviewed (approved):
LGTM, thanks @MaiRajborirug!
